### ISSUE TITLE
Jruby 9k raises exceptions in the wrong spot (instability)

### ISSUE THREAD

**Issue by andrewvc** [elastic/logstash#8008]
In some situations I've noticed that raising an exception in a plugin can cause Jruby to die with an exception coming from the wrong place, thus dumping a bad stack trace. Usually that stacktrace is something like:

```
[2017-08-15T21:48:40,376][FATAL][logstash.runner          ] An unexpected error occurred! {:error=>#<NoMethodError: undefined method `[]=' for #<LogStash::OutputDelegator:0x24ae124e>>, :backtrace=>["/Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/pipeline.rb:519:in `block in output_batch'", "org/jruby/RubyArray.java:1734:in `each'", "/Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/pipeline.rb:526:in `block in output_batch'", "/Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:227:in `block in each'", "/Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:226:in `each'", "/Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/pipeline.rb:520:in `output_batch'", "/Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/pipeline.rb:480:in `worker_loop'", "/Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/pipeline.rb:437:in `block in start_workers'"]}
```

Sometimes it escapes JRuby's regular stack traces entirely like:

```

Unhandled Java exception: java.lang.BootstrapMethodError: java.lang.NoSuchMethodError: Users.andrewcholakian.projects.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$block$output_batch$3(Lorg/jruby/runtime/ThreadContext;Lorg/jruby/runtime/Block;Lorg/jruby/parser/StaticScope;Lorg/jruby/runtime/builtin/IRubyObject;[Lorg/jruby/runtime/builtin/IRubyObject;Lorg/jruby/runtime/Block;Ljava/lang/String;Lorg/jruby/runtime/Block$Type;)Lorg/jruby/runtime/builtin/IRubyObject;
java.lang.BootstrapMethodError: java.lang.NoSuchMethodError: Users.andrewcholakian.projects.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$block$output_batch$3(Lorg/jruby/runtime/ThreadContext;Lorg/jruby/runtime/Block;Lorg/jruby/parser/StaticScope;Lorg/jruby/runtime/builtin/IRubyObject;[Lorg/jruby/runtime/builtin/IRubyObject;Lorg/jruby/runtime/Block;Ljava/lang/String;Lorg/jruby/runtime/Block$Type;)Lorg/jruby/runtime/builtin/IRubyObject;
           block in output_batch at /Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/pipeline.rb:526
                      callDirect at org/jruby/runtime/CompiledIRBlockBody.java:145
                      callDirect at org/jruby/runtime/MixedModeIRBlockBody.java:113
                            call at org/jruby/runtime/IRBlockBody.java:71
                            call at org/jruby/runtime/Block.java:124
                            call at org/jruby/RubyProc.java:289
                          call19 at org/jruby/RubyProc.java:273
                   block in each at /Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:227
                     yieldDirect at org/jruby/runtime/CompiledIRBlockBody.java:156
                           yield at org/jruby/runtime/BlockBody.java:114
                           yield at org/jruby/runtime/Block.java:165
                            each at org/jruby/javasupport/ext/JavaLang.java:93
                            each at org/jruby/javasupport/ext/JavaUtil.java:133
                            each at /Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:226
  RUBY$method$each$0$__VARARGS__ at Users/andrewcholakian/projects/logstash/logstash_minus_core/lib/logstash/util//Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:-1
                            call at org/jruby/internal/runtime/methods/CompiledIRMethod.java:77
                            call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:93
                       callBlock at org/jruby/runtime/callsite/CachingCallSite.java:77
                            call at org/jruby/runtime/callsite/CachingCallSite.java:83
                       interpret at org/jruby/ir/instructions/CallBase.java:428
                     processCall at org/jruby/ir/interpreter/InterpreterEngine.java:355
                       interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:73
                       interpret at org/jruby/ir/interpreter/InterpreterEngine.java:83
                INTERPRET_METHOD at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:179
                            call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:165
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:200
                    cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:338
                            call at org/jruby/runtime/callsite/CachingCallSite.java:163
                     processCall at org/jruby/ir/interpreter/InterpreterEngine.java:314
                       interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:73
                       interpret at org/jruby/ir/interpreter/InterpreterEngine.java:89
                INTERPRET_METHOD at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:214
                            call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:200
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:208
          block in start_workers at /Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/pipeline.rb:437
                      callDirect at org/jruby/runtime/CompiledIRBlockBody.java:145
                            call at org/jruby/runtime/IRBlockBody.java:71
                            call at org/jruby/runtime/Block.java:124
                            call at org/jruby/RubyProc.java:289
                            call at org/jruby/RubyProc.java:246
                             run at org/jruby/internal/runtime/RubyRunnable.java:104
                             run at java/lang/Thread.java:748

Caused by:
java.lang.NoSuchMethodError: Users.andrewcholakian.projects.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$block$output_batch$3(Lorg/jruby/runtime/ThreadContext;Lorg/jruby/runtime/Block;Lorg/jruby/parser/StaticScope;Lorg/jruby/runtime/builtin/IRubyObject;[Lorg/jruby/runtime/builtin/IRubyObject;Lorg/jruby/runtime/Block;Ljava/lang/String;Lorg/jruby/runtime/Block$Type;)Lorg/jruby/runtime/builtin/IRubyObject;
                         resolve at java/lang/invoke/MethodHandleNatives.java:-2
                         resolve at java/lang/invoke/MemberName.java:975
                   resolveOrFail at java/lang/invoke/MemberName.java:1000
                   resolveOrFail at java/lang/invoke/MethodHandles.java:1394
        linkMethodHandleConstant at java/lang/invoke/MethodHandles.java:1750
        linkMethodHandleConstant at java/lang/invoke/MethodHandleNatives.java:477
           block in output_batch at /Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/pipeline.rb:526
                      callDirect at org/jruby/runtime/CompiledIRBlockBody.java:145
                      callDirect at org/jruby/runtime/MixedModeIRBlockBody.java:113
                            call at org/jruby/runtime/IRBlockBody.java:71
                            call at org/jruby/runtime/Block.java:124
                            call at org/jruby/RubyProc.java:289
                          call19 at org/jruby/RubyProc.java:273
                   block in each at /Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:227
                     yieldDirect at org/jruby/runtime/CompiledIRBlockBody.java:156
                           yield at org/jruby/runtime/BlockBody.java:114
                           yield at org/jruby/runtime/Block.java:165
                            each at org/jruby/javasupport/ext/JavaLang.java:93
                            each at org/jruby/javasupport/ext/JavaUtil.java:133
                            each at /Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:226
  RUBY$method$each$0$__VARARGS__ at Users/andrewcholakian/projects/logstash/logstash_minus_core/lib/logstash/util//Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:-1
                            call at org/jruby/internal/runtime/methods/CompiledIRMethod.java:77
                            call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:93
                       callBlock at org/jruby/runtime/callsite/CachingCallSite.java:77
                            call at org/jruby/runtime/callsite/CachingCallSite.java:83
                       interpret at org/jruby/ir/instructions/CallBase.java:428
                     processCall at org/jruby/ir/interpreter/InterpreterEngine.java:355
                       interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:73
                       interpret at org/jruby/ir/interpreter/InterpreterEngine.java:83
                INTERPRET_METHOD at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:179
                            call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:165
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:200
                    cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:338
                            call at org/jruby/runtime/callsite/CachingCallSite.java:163
                     processCall at org/jruby/ir/interpreter/InterpreterEngine.java:314
                       interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:73
                       interpret at org/jruby/ir/interpreter/InterpreterEngine.java:89
                INTERPRET_METHOD at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:214
                            call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:200
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:208
          block in start_workers at /Users/andrewcholakian/projects/logstash/logstash-core/lib/logstash/pipeline.rb:437
                      callDirect at org/jruby/runtime/CompiledIRBlockBody.java:145
                            call at org/jruby/runtime/IRBlockBody.java:71
                            call at org/jruby/runtime/Block.java:124
                            call at org/jruby/RubyProc.java:289
                            call at org/jruby/RubyProc.java:246
                             run at org/jruby/internal/runtime/RubyRunnable.java:104
                             run at java/lang/Thread.java:748
```

To reproduce this simply run master with this bad branch of of logstash-integration-internal https://github.com/andrewvc/logstash-integration-internal/tree/bad-situation . 

This will error on startup causing those errors a significant amount of time.

Use the following `logstash/pipelines.yml`:
```
- pipeline.id: publisher
  config.string: "input { generator { message => huh } } output { internal { send_to => [foo] } }"
- pipeline.id: publishertoo
  config.string: "input { generator { message => whut } } output { internal { send_to => [foo] } }"
- pipeline.id: subscriber
  config.string: "input { internal { address => foo } } output { stdout {} }"
```

**Comment by headius** [comment]
This likely also stems from the issues in jruby/jruby#4739.

**Comment by andrewvc** [comment]
This should be fixed in #8144
